### ISSUE TITLE
Headless Timeskipping Causing Video Jumping

### ISSUE THREAD

**Issue by Enverex** [Neos-Metaverse/NeosPublic#2943]
### Describe the bug?

While the title says video skipping, I'm sure there would be other side effects that I've just personally not seen.

Ever since the headless time stream addition (on 2021.8.29.1240) I've now noticed videos keep skipping backwards every so often. Normally it affects everyone the same amount, but sometimes a few users will avoid it and others will be offset by a slightly different amount.

### Relevant issues

None found.

### To Reproduce

I'm not sure on the best reproduction environment for this, but a 100% sure way to reproduce this is:

Join the Cineworld session, go to Screen 2, gather some people (more people seems to result in it happening more reliably, not sure if they are the cause), play a 20 minute or so video of your choice. Technically playing movies in the other screens should work too but this is faster.

### Expected behavior

For videos to play smoothly.

### Log Files

Nothing is logged client or server side when this happens.

### Screenshots

A few examples of it being captured. Added a UVO later on to see the per-user values.

![image](https://user-images.githubusercontent.com/958670/132137786-18b861be-6053-4292-b2ad-f0f35fef4c83.png)

![image](https://user-images.githubusercontent.com/958670/132137793-fb4ad678-6693-4c3b-93dc-c4d51057e2f6.png)

![image](https://user-images.githubusercontent.com/958670/132137811-c553583e-e9d6-40c1-be8f-5058011fc67b.png)

### How often does it happen?

Always

### Does the bug persist after restarting Neos?

Yes

### Neos Version Number

2021.9.4.468

### What Platforms does this occur on?

Windows, Linux

### Link to Reproduction Item/World

http://cloudx.azurewebsites.net/open/session/S-U-EnverexBot:Cineworld

### Did this work before?

Yes

### If it worked before, on which build?

2021.8.29.1240

**Comment by kulzae** [comment]
Also having this occur on my headless. though the amount of time skipped is usually around a second so its much less noticeable. im not sure what factors are contributing to the time diffrence. Possibly length of video?

**Comment by Lexevolution** [comment]
It doesn't happen too regularly when on my friend's headless but when it does, it's usually by a couple of seconds. Happens seemingly randomly though so I'm not 100% sure how to diagnose it.

**Comment by Frooxius** [comment]
- If I understand the LogiX correctly, it's essentially showing that the time has jumped back around 10 seconds for most people? Did you get those numbers after the users were actually joined in?

The reason I ask is because you can get a big timejump like this when the user initially joins, so I want to make sure that's not what we're seeing in the numbers.

- How often (roughly) does this happen in a session? What's the typical interval between the jumps?

- Is this correlated with any event? E.g. something lagging out or someone joining the session? Possibly any correlation to CPU/Memory usage of the headless?

**Comment by Enverex** [comment]
>If I understand the LogiX correctly, it's essentially showing that the time has jumped back around 10 seconds for most people?

Yes. The amount of time it jumps back doesn't appear to be fixed but it seems most often around 10 seconds.

>Did you get those numbers after the users were actually joined in? The reason I ask is because you can get a big time-jump like this when the user initially joins, so I want to make sure that's not what we're seeing in the numbers.

I initially suspected the same thing, but the time jumps continued happening even after no-one had joined or left for some time.

>How often (roughly) does this happen in a session? What's the typical interval between the jumps?

There doesn't appear to be any obvious repeating period of time. Sometimes it will be minutes apart, other times it would happen several times a minute. There was one instance where it got a little out of control and was skipping both backwards AND forwards a few seconds, every few seconds (everyone saw this, not just one person) - this stopped after about 30 seconds.

>Is this correlated with any event? E.g. something lagging out or someone joining the session? Possibly any correlation to CPU/Memory usage of the headless?

The headless wasn't stressed when this was happening. Neos was using around 80% of one core of an 8 core machine (8 physical, 16 logical cores). It had around 55GB of RAM still free as well. There did not appear to be any CPU spikes on the server to indicate Neos was suddenly struggling for any reason.

**Comment by Frooxius** [comment]
Thanks for the info. I've made some tweaks and added some extra diagnostics in 2021.9.6.614, please let me know if it helps or send new logs if the issue still persists!

**Comment by Enverex** [comment]
Still appears to be happening. It happens every time someone leaves (which didn't used to be the case before the original headless changes were made) but the skips unrelated to joins still appear to be happening. From what I can tell, they get worse the longer the session has been open (or the longer you've been in the session?).

Here are logs from a few people in the same session:
[ENVEREX - 2021.9.6.614 - 2021-09-07 19_19_39.log.zip](https://github.com/Neos-Metaverse/NeosPublic/files/7124076/ENVEREX.-.2021.9.6.614.-.2021-09-07.19_19_39.log.zip)
[MATTYK - 2021.9.6.614 - 2021-09-07 16_48_32.log.zip](https://github.com/Neos-Metaverse/NeosPublic/files/7124077/MATTYK.-.2021.9.6.614.-.2021-09-07.16_48_32.log.zip)
[TAKO - 2021.9.6.614 - 2021-09-07 20_40_31.log.zip](https://github.com/Neos-Metaverse/NeosPublic/files/7124078/TAKO.-.2021.9.6.614.-.2021-09-07.20_40_31.log.zip)

**Comment by Ducky07** [comment]
Adding mine to the list as well.

[Napo - 2021.9.6.614 - 2021-09-07 21_53_17.zip](https://github.com/Neos-Metaverse/NeosPublic/files/7124217/Napo.-.2021.9.6.614.-.2021-09-07.21_53_17.zip)

**Comment by Frooxius** [comment]
Thanks for the logs! I reworked a bunch of things in 2021.9.8.946, please let me know if you notice any improvements!

If the issue still happens, can you send me new logs please? Also if you can, there's a new EstimatedMasterClockError LogiX node, can you watch what happens with it before and after the time skip (especially before though)? Especially I'd be interestd to see if it shows really large offset that keeps growing over time.

**Comment by Enverex** [comment]
So... mixed results.

The good news: Time skipping appears to have been fixed, at least it's not doing it randomly and didn't seem to jump when people joined/left, at least not in preliminary testing.

The bad news: Playback positions seem to be wrong for people that join while playback is in progress. Oddly enough the playback position is ahead of where it should be rather than behind. I was seeing a playback position around 30 minutes ahead of where someone else who'd been in the world much longer was seeing.

Log of the session where I was desynced (way ahead of time):
[JASPER - 2021.9.8.946 - 2021-09-08 18_13_46.log](https://github.com/Neos-Metaverse/NeosPublic/files/7131126/JASPER.-.2021.9.8.946.-.2021-09-08.18_13_46.log)

Note that anything that was started while I was there worked fine.

**Comment by Frooxius** [comment]
Hmm such large drift is very bizarre, there must be some wrong offset being calculated somewhere.

Can you check what the world Time (e.g. from the LogiX T node) was for you while this was happening and what it was for others?

What did the EstimatedMasterClockError say?

**Comment by Enverex** [comment]
EstimatedMasterClockError (for me) was very low - 0.00(something).
Unfortunately everyone's left now so I can't check their T values.

**Comment by Frooxius** [comment]
Hmm, would you be able to check it next time?

For the video, how was it being played back?

Also was the position slider for the playback also significantly further ahead of time or was it roughly in the correct position to what it should've been?

**Comment by Enverex** [comment]
Two videos were going on in the world when I joined. The first one for me wasn't even playing as it had "finished" (the two people watching were synced with each other fine and still seeing it). The second video elsewhere was way ahead for me. 

Basically playback was fine, but I appeared to be offset and video players even showed this offset as if it was correct. e.g. the first video was showing as stopped due to completing. The second video I could see the playback bar almost at the very end (nearly finished) while those that had been there longer saw the playback bar in what I believe was the actual correct location, which was around 2/3rds the way through.

In short, playback positions in video players mirrored what each user saw (even if it was wrong).

I'll try and get T values next time I can get people together or can find users split up enough to recreate it.

**Comment by Frooxius** [comment]
I see thanks! It seems like there might've been a pretty big offset for some reason. Checking the T would definitely help here, especially if it's significantly different to other users.

**Comment by Enverex** [comment]
Ok, so everyone's T value matched within a second or two at most. EstimatedMasterClockError as very low for everyone (everyone below 0.0).

What was observed, was that anyone that joined while a video was already playing, appeared to have an obscene offset on the video itself.

I was in the world when the video started, so everything is fine.
![image](https://user-images.githubusercontent.com/958670/132589298-23a9146c-c901-47c1-b44d-d196a5786217.png)

Both people that joined after the video had already been running for a few minutes had this result.
![image](https://user-images.githubusercontent.com/958670/132589419-96bc7725-249b-419f-823b-ca1a88d05ae1.png)

The clock error appears to be roughly the same length as the video itself (which is just under 131 minutes).

**Comment by Frooxius** [comment]
Ah thanks for the info! I know what it is now, got a fix coming soon!

**Comment by Frooxius** [comment]
Ok, released the fix in 2021.9.9.1144. This seems to have been caused by the rework actually, the initial session time wouldn't get synchronized until after the initial world state was decoded, which resulted in it having wildly wrong offset.

It should be gone now, I'm going to close this since time sync seems to be good otherwise. Please reopen if there are more issues.
